### PR TITLE
Digital signal pulse width

### PR DESCRIPTION
**Pull request by m-nolan**
#257 Add code fulfilling issue 157 requirements

### CODE REVIEW

File: aopy/utils.py [hunk 1]
```diff
@@ -300,6 +300,41 @@ def get_edges_from_onsets(onsets, pulse_width):
         values[2+2*t] = 0
     return timestamps, values
 
+def get_pulse_times( digital_data, sync_channel, trigger_channel, samplerate ):
```
Comment (leoscholl) [inline aopy/utils.py:303]:
```suggestion
def get_e3v_frame_timestamps( digital_data, sync_channel, trigger_channel, samplerate ):
```

**Review by leoscholl — COMMENTED** [review]
i think since this is now specific to e3v cameras (sync and trigger channels, 6 pulses, etc.), it is no longer a "utility" and more of a preprocessing function. i would rename it and maybe move it somewhere else. maybe separate the calculating pulse times and duty cycle into its own function?

**Comment by m-nolan** [comment]
refactored the code as asked.

File: aopy/data.py [hunk 2]
```diff
@@ -1,7 +1,11 @@
 # data.py
 # Code for directly loading and saving data (and results)
 
+from ctypes import util
+from random import sample
 import numpy as np
+
+from aopy import utils
```
Comment (leoscholl) [inline aopy/data.py:8]: this doesn't look right. should be relative import `from . import utils`

File: aopy/data.py [hunk 3]
```diff
@@ -1069,3 +1074,34 @@ def yaml_read(filename):
...
+
+
+def get_e3v_video_frame_data( digital_data, sync_channel_idx, trigger_channel_idx, samplerate ):
+    
+    """get_e3v_video_frame_data
+
+    Compute pulse times and duty cycles from e3vision video data frames collected on an ecube digital panel.
+
+    Args:
+        digital_data (np.array): [n_time x n_channel] array of data read from ecube digital panel
```
Comment (leoscholl) [inline aopy/data.py:1086]: i prefer 
```suggestion
        digital_data (nt, nch): array of data read from ecube digital panel
```

File: aopy/utils.py [hunk 4]
```diff
@@ -300,6 +300,44 @@ def get_edges_from_onsets(onsets, pulse_width):
         values[2+2*t] = 0
     return timestamps, values
 
+def get_pulse_edge_times( digital_data, samplerate ):
+
+    """get_pulse_edge_times
+
+    Args:
+        digital_data (np.array): [n_time x 1] array of data from ecube digital panel
```
Comment (leoscholl) [inline aopy/utils.py:308]: same formatting as above

File: aopy/utils.py [hunk 5]
```diff
@@ -300,6 +300,44 @@ def get_edges_from_onsets(onsets, pulse_width):
...
+
+def compute_pulse_duty_cycles( edge_pairs ):
+
+    """compute_pulse_duty_cycles
+
+    Args:
+        edge_pairs (np.array): [n_pulse x 2] start, end times from a series of pulses
+
+    Returns:
+        duty_cycle (): _description_
```
Comment (leoscholl) [inline aopy/utils.py:332]: finish this docstring

File: aopy/data.py [hunk 6]
```diff
@@ -1069,3 +1074,34 @@ def yaml_read(filename):
...
+
+    Returns:
+        pulse_times (np.array): array of floats indicating pulse start times
+        duty_cycle (np.array): array of floats indicating pulse duty cycle (quotient of pulse width and pulse period)
+    """
+
+    trig_pulse_edges = get_pulse_edge_times(digital_data[:,trigger_channel_idx],samplerate)
+    trig_pulse_times = trig_pulse_edges[:,0]
+    start_trig_time = trig_pulse_times[0]
+    end_trig_time = trig_pulse_times[3]
```
Comment (leoscholl) [inline aopy/data.py:1099]: a comment here about what the trigger signal looks like might be useful for people looking at this in the future

**Review by leoscholl — APPROVED** [review]
cool, just a few minor comments. also technically compute_pulse_duty_cycles( ) and get_pulse_edge_times() should have tests of their own. maybe write something simple. or maybe write an empty test and point people to the test for get_e3v_video_frame_data()

File: aopy/data.py (hunk 3 above)
Comment (m-nolan) [reply aopy/data.py:1086]: the format is `name (type): description`, right?

Comment (leoscholl) [reply aopy/data.py:1086]: technically it's `varIn_name1 (dim label1, dim label2): Description`
see https://analyze.readthedocs.io/en/latest/documentation-guidelines.html

**Pull request merged**